### ISSUE TITLE
parking_lot mutex is slower than standard library

### ISSUE THREAD

**Issue by tsxiaofang** [tokio-rs/tokio#6317]
I also found this issue during performance testing, and I think it's best to disable parking_lot feature by default.
https://github.com/Amanieu/parking_lot/issues/338

**Comment by carllerche** [comment]
Thanks for the report. This is interesting. I think before making a decision on what to do with this, we need more data. How much does this impact Tokio in practice?

Assuming the issue does impact Tokio, we have a few options. Technically, we don't *have* to actually use the parking lot mutexes when the feature is enabled. It looks like the issue only impacts some platforms. One option would be to conditionally compile using the "best" mutex based on the platform. If some end users want strict control over which mutex implementation Tokio uses, we could provide `cfg` flags for the end application to use. E.g. `--cfg tokio_force_parking_lot_mutex` (or something).

**Comment by tsxiaofang** [comment]
Yes, my current approach is to disable the default features and add the necessary features myself

**Comment by madsager** [comment]
We encountered this as well. We depend on other crates that use tokio with the `full` feature. Since that includes `parking_lot` we had to do some build hacking to disable `parking_lot`. The performance difference is significant. It would be great if the `full` feature did not imply `parking_lot` on the platforms where it does not perform well. Is work planned in this area? It would be nice to get full performance from tokio without having to do build hacking. :)

**Comment by Darksonn** [comment]
It used to be the case that some methods were only available when `parking_lot` was enabled, but I don't think the parking_lot has any observable impact on the behavior of Tokio anymore. It might be ok to just drop it from `full`.
